pkg/apiserver/user/sqlauth/sqlauth.go

```diff
@@ -50,13 +67,123 @@ func (a *Authenticator) Authenticate(f user.AuthenticateForm) (*utils.SessionUse
 	}
 	defer utils.CloseTiDBConnection(db) //nolint:errcheck
 
+	// Check dashboard privileges
+	// 1. Check whether TiDB SEM is enabled
+	resData, err := a.tidbClient.SendGetRequest("/config")
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	var config tidbSecurityConfig
+	err = json.Unmarshal(resData, &config)
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	// 2. Get grants
+	var grantRows []string
+	err = db.Raw("show grants for current_user()").Find(&grantRows).Error
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	grants := parseCurUserGrants(grantRows)
+	// 3. Check
+	if !checkDashboardPriv(grants, config.Security.EnableSem) {
+		return nil, user.ErrLackPrivileges.NewWithNoMessage()
+	}
+
 	return &utils.SessionUser{
 		Version:      utils.SessionVersion,
 		HasTiDBAuth:  true,
 		TiDBUsername: f.Username,
 		TiDBPassword: f.Password,
 		DisplayName:  f.Username,
 		IsShareable:  true,
-		IsWriteable:  true,
+		IsWriteable:  checkWriteablePriv(grants),
 	}, nil
 }
+
+// Currently, There are 2 kinds of grant output format in TiDB:
+// - GRANT [grants] ON [db.table] TO [user]
+// - GRANT [roles] TO [user]
+// Examples:
+// - GRANT PROCESS,SHOW DATABASES,CONFIG ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT SYSTEM_VARIABLES_ADMIN,RESTRICTED_VARIABLES_ADMIN,RESTRICTED_STATUS_ADMIN,RESTRICTED_TABLES_ADMIN ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT ALL PRIVILEGES ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT `app_read`@`%` TO `test`@`%`
+func parseCurUserGrants(grantRows []string) set {
+	grants := set{}
+	grantRegex := regexp.MustCompile(`GRANT (.+) ON`)
```

Comment:
It is recommended to move this `MustCompile` to the outside, as a variable, so that:

1. It is not repeatedly compiled.
2. An regex mistake can be discovered early, i.e. resulting in a panic at TiDB Dashboard starting phase, instead of panic when `parseCurUserGrants` is called.

Reply:
Done.